File: RUBY/Cap08/c08ex07.rb
# c08ex07.rb

#!/usr/bin/env ruby

LIMITE = 10
$matriz = Array.new(LIMITE)

def vazia
  if ($topo == -1)
    true
  else
    false
  end
end

def cheia
  if ($topo == LIMITE - 1)
    true
  else
    false
  end
end

def adicionar(elemento)
  if (cheia)
    false
  else
    $topo += 1
    $matriz[$topo] = elemento
    true
  end
end

def retirar(elemento)
  if (vazia)
    false
  else
    elemento[0] = $matriz[$topo]
    $topo -= 1
    true
  end
end

def empilhar
  print "Entre um elemento numerico: "
  x = gets.chomp.to_i
  puts "\n"
  if (adicionar(x))
    puts "Elemento #{x} inserido na posiçao #{$topo}."
  else
    puts "Impossivel adicionar #{x} - pilha cheia.\n"
  end
  puts
end

def desempilhar
  if (vazia)
    puts "Impossivel retirar elemento - pilha vazia.\n"
  else
    elemento = Array.new(1)
    if (retirar(elemento))
      puts "Elemento #{elemento[0]} removido do topo da pilha.\n"
    end
  end
  puts
end

def mostrar
  if (!(vazia))
    for i in $topo.downto(0)
      printf "Posicao %2d: %d\n", i+1, $matriz[i]
    end
  else
    puts "Impossivel mostrar - pilha vazia.\n"
  end
  puts
end

def criar
  $topo = -1
  for i in 1..(LIMITE - 1)
    $matriz[i] = 0
  end
end

criar
opcao = 0
while (opcao != 5)
  puts "PROGRAMA: PILHA\n\n"
  puts "[1] - Empilhar\n"
  puts "[2] - Desempilhar\n"
  puts "[3] - mostrar\n"
  puts "[4] - Criar pilha\n"
  puts "[5] - Sair\n\n"

  print "Escolha uma opcao: "
  opcao = gets.chomp.to_i
  puts "\n"

  if (opcao != 5)
    case (opcao)
    when 1
      empilhar
    when 2
      desempilhar
    when 3
      mostrar
    when 4
      criar
    end
  end
end

File: LUA/Cap08/c08ex07.lua
-- c08ex07.lua

LIMITE = 10
matriz = {}

local function vazia()
  if (topo == 0) then
    return true
  else
    return false
  end
end

local function cheia()
  if (topo == LIMITE) then
    return true
  else
    return false
  end
end

local function adicionar(elemento)
  if (cheia()) then
    return false
  else
    topo = topo + 1
    matriz[topo] = elemento
    return true
  end
end

local function retirar(elemento)
  if (vazia()) then
    return false
  else
    elemento[1] = matriz[topo]
    topo = topo - 1
    return true
  end
end

local function empilhar()
  io.write("Entre um elemento numerico: ")
  local x = tonumber(io.read())
  print()
  if (adicionar(x)) then
    print("Elemento " .. x .. " inserido na posicao " .. topo .. ".")
  else
    print("Impossivel adicionar " .. x .. " - pilha cheia.")
  end
  print()
end

local function desempilhar()
  if (vazia()) then
    print("Impossivel retirar elemento - pilha vazia.")
  else
    local elemento = {1}
    if (retirar(elemento)) then
      print("Elemento " .. elemento[1] .. " removido do topo da pilha.")
    end
  end
  print()
end

local function mostrar()
  if (not vazia()) then
    for i = topo, 1, -1 do
      print("Posicao " .. i .. ": " .. matriz[i])
    end
  else
    print("Impossivel mostrar - pilha vazia.")
  end
  print()
end

local function criar()
  topo = 0
  for i = 1, LIMITE do
    matriz[i] = 0
  end
end

criar()
local opcao = 0
while (opcao ~= 5) do
  print("PROGRAMA: PILHA\n")
  print("[1] - Empilhar")
  print("[2] - Desempilhar")
  print("[3] - mostrar")
  print("[4] - Criar pilha")
  print("[5] - Sair\n")

  io.write("Escolha uma opcao: ")
  opcao = tonumber(io.read())
  print()

  if (opcao ~= 5) then
    if (opcao == 1) then
      empilhar()
    elseif (opcao == 2) then
      desempilhar()
    elseif (opcao == 3) then
      mostrar()
    elseif (opcao == 4) then
      criar()
    end
  end
end
